src/ca/verify/signer.ts

```diff
@@ -0,0 +1,136 @@
+
+// Verifies the identity embedded in a Fulcio-issued signing certificate against
+// the list of trusted identities. Returns without error if at least one of the
+// identities matches the signing certificate; otherwise, throws a
+// VerificationError.
+export function verifySignerIdentity(
+  signingCert: x509Certificate,
+  identities: sigstore.CertificateIdentities
+): void {
+  // Check that the signing certificate was issued to at least one of the
+  // specified identities
+  const signerVerified = identities.identities.some((identity) =>
+    verifyIdentity(signingCert, identity)
+  );
+
+  if (!signerVerified) {
+    throw new VerificationError('Certificate issued to untrusted signer');
+  }
+}
+
+// Checks that the specified certificate was issued to the specified identity.
+// The certificate must match the issuer, subject alternative name, and an
+// optional list of certificate extensions. Returns true if the certificate was
+// issued to the identity; otherwise, returns false.
+function verifyIdentity(
+  cert: x509Certificate,
+  identity: sigstore.CertificateIdentity
+): boolean {
+  return (
+    verifyIssuer(cert, identity.issuer) &&
+    verifySAN(cert, identity.san) &&
+    verifyOIDs(cert, identity.oids)
+  );
+}
+
+// Checks the Fulcio issuer extension against the expected issuer. Returns true
+// if the issuer matches; otherwise, returns false.
+function verifyIssuer(cert: x509Certificate, issuer: string): boolean {
+  const issuerExtension = cert.extension(OID_FUCIO_ISSUER);
+
+  return issuerExtension?.value.toString('ascii') === issuer;
+}
+
+// Checks the certificate against the expected subject alternative name. Returns
+// true if the SAN matches; otherwise, returns false.
+function verifySAN(
+  cert: x509Certificate,
+  expectedSAN?: sigstore.SubjectAlternativeName
+): boolean {
+  // Fail if the SAN is not specified or is not a supported type
+  if (
+    expectedSAN === undefined ||
+    expectedSAN.identity === undefined ||
+    expectedSAN.type ===
+      sigstore.SubjectAlternativeNameType
+        .SUBJECT_ALTERNATIVE_NAME_TYPE_UNSPECIFIED
+  ) {
+    return false;
+  }
+
+  const sanExtension = cert.extSubjectAltName;
+
+  // Fail if the certificate does not have a SAN extension
+  if (!sanExtension) {
+    return false;
+  }
+
+  let sanValue: string | undefined;
+  switch (expectedSAN.type) {
+    case sigstore.SubjectAlternativeNameType.EMAIL:
+      sanValue = sanExtension.rfc822Name;
+      break;
+    case sigstore.SubjectAlternativeNameType.URI:
+      sanValue = sanExtension.uri;
+      break;
+    case sigstore.SubjectAlternativeNameType.OTHER_NAME:
+    // TODO: Support other name types
+  }
+
+  // Missing SAN value is an automatic failure
+  if (sanValue === undefined) {
+    return false;
+  }
+
+  let match: string | undefined;
+  switch (expectedSAN.identity.$case) {
+    case 'value':
+      match = expectedSAN.identity.value;
+      break;
+    case 'regexp':
+      // TODO support regex
```

Comment:
I have lots of questions about how this is supposed to work. Gonna circle back on supporting regex matches for SAN values.

Reply:
Good call! Seems like it should really be a strict subset of regex, like only allow simple globs?

Reply:
Love that we're implementing this so close to the specs and raising these questions.